pennylane/gradients/parameter_shift.py

```diff
@@ -140,34 +140,72 @@ def _extract_unshifted(recipe, at_least_one_unshifted, f0, gradient_tapes, tape)
     return recipe, at_least_one_unshifted, unshifted_coeff
 
 
-def _evaluate_gradient(res, data, broadcast, r0, scalar_qfunc_output):
+def _unshifted_coeff(g, unshifted_coeff, r0):
+    """Auxiliary function; if unshifted term exists, add its contribution."""
+    if unshifted_coeff is not None:
+        # add the unshifted term
+        g = g + unshifted_coeff * r0
+    return g
+
+
+def _evaluate_gradient(tape, res, data, broadcast, r0, scalar_qfunc_output):
     """Use shifted tape evaluations and parameter-shift rule coefficients
     to evaluate a gradient result."""
-
     _, coeffs, fn, unshifted_coeff, batch_size = data
 
     # individual post-processing of e.g. Hamiltonian grad tapes
     if fn is not None:
         res = fn(res)
 
-    # compute the linear combination of results and coefficients
-    axis = 0
-    if not broadcast:
-        res = qml.math.stack(res)
-    elif (
-        qml.math.get_interface(res[0]) != "torch"
-        and batch_size is not None
-        and not scalar_qfunc_output
-    ):
-        # If the original output is not scalar and broadcasting is used, the second axis
-        # (index 1) needs to be contracted. For Torch, this is not true because the
-        # output of the broadcasted tape is flat due to the behaviour of the Torch device.
-        axis = 1
-    g = qml.math.tensordot(res, qml.math.convert_like(coeffs, res), [[axis], [0]])
+    if not qml.active_return():

```

Comment:
I found it cleaner to separate the functions by adding `_new`, the question here is do we consider that gradient_transform are user facing? If yes then indeed it is better to keep only one function otherwise I found the other better, because users will simply use `qml.jacobian` or `qml.grad` with the keyword `"param-shift"` in the QNode when the integration is done.

Reply:
> users will simply use qml.jacobian or qml.grad with the keyword "param-shift" in the QNode when the integration is done.

Not sure I follow. How is this related to either having `_new` functions or `qml.active_return()` branching logic? To me, it seems users would have to do `qml.enable_return()` either way.